[2409] Csv Spawner Notification Bus

w-czerski opened the pull request:
## About:
This PR introduces a notification bus for the CSV spawner.
It provides the ability to trigger actions when entities are spawned and to gather information when the spawning process begins.
This is particularly useful when there is a need to react to the moment _"when entities are spawned"_.

w-czerski committed:
init spline subscriber

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix | get spline and convert into vector of vertex

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove auto

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
apply review changes

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove comments

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
use emplace | format | apply reviews

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
init | add notification bus interface class

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

arturkamieniecki reviewed (commented):
There is a lot of info passed in the bus calls. Verify if everything is needed. Consider creating a struct for this data.

Overall looks good!

w-czerski committed:
remove not needed info in bus

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove bus info for spawn tickets

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

patrykantosz reviewed (commented):
Code looks okay. I'm just curious if anything (beyond this PR, as it is obvious the implementation is not here) implements handlers for these two newly added functions? Or has the bus been created for some further work planned for the future?

w-czerski commented:
In current state, we need some explicit logic to handle is terrain created (please look at #102). 
This is just a helpful functionality to prevent users to stacks tick frames, or queue functions. User can directly do something on spawn ready, and skip in most cases terrain handling (since this is not useful or needed for user).

Exactly this functionality is needed in one project I'm working on. I would reconsider what information should be passed down in the buses. Right now notification is only needed.

w-czerski committed:
move spawn utils to public include | spawn info bus struct

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add notify status code

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix typo

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
more context to enum spawn

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix format | resolve errors

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
rename member

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add default override implentation | make override optional

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add description

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski commented:
I've added more context, introduced struct with info and result status code. 
Let me know what you think. I've made it more generic for the future.

jhanca-robotecai commented on Gems/CsvSpawner/Code/Include/CsvSpawner/CsvSpawnerInterface.h:
```diff
@@ -0,0 +1,86 @@
+     */
+    enum class SpawnStatusCode : uint8_t
+    {
+        Success = 0, ///< Operation succeeded.
+        Fail = 1 << 0, ///< Generic failure.
+        SpawnStopped = 1 << 1, ///< Spawning was stopped prematurely but not necessarily a failure.
+        ErrorOccurred = 1 << 2, ///< An error occurred during spawning (potentially recoverable).
```
Are the bitshifts needed here? Typically, the bitshifts are used when multiple flags can happen at once.

w-czerski replied:
Overall, I thought - spawn can still be successful even when some errors / issues occur. Like partialy some entites are spawned, some are spawned but not in desired place etc.

jhanca-robotecai commented on Gems/CsvSpawner/Code/Include/CsvSpawner/CsvSpawnerInterface.h:
```diff
@@ -0,0 +1,86 @@
+         * @param m_spawnInfo Struct holding information about entities to be spawned.
+         * @param m_statusCode Status code indicating success, failure and warnings of the spawn.
+         */
+        virtual void OnEntitiesSpawnFinished(const SpawnInfo& m_spawnInfo, const SpawnStatusCode& m_statusCode)
```
`const` has no use in definitions, hence it should be used only to increase the readability of the code. `m_statusCode` *indicates success, failure and warnings*, but the `const` indicator would suggest the value will not be changed by the method, hence the indication will not take any place.

jhanca-robotecai commented on Gems/CsvSpawner/Code/Include/CsvSpawner/CsvSpawnerInterface.h:
```diff
@@ -0,0 +1,86 @@
+     * CsvSpawnerInterface is an Event Bus interface that notifies multiple
+     * listeners when entity spawning begins and finishes.
+     */
+    class CsvSpawnerInterface : public AZ::EBusTraits
```
Shouldn't the interface method stay purely virtual?

w-czerski replied:
Making it all pure virtual forces the user to implement an override for both functions.

jhanca-robotecai commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -196,6 +196,13 @@ namespace CsvSpawner::CsvSpawnerUtils
     {
+        SpawnInfo broadcastSpawnInfo =
+            SpawnInfo{ entitiesToSpawn, physicsSceneName, parentId }; // Spawn Info used in CsvSpawner EBus notify.
+        SpawnStatusCode spawnStatusCode; // Spawn Status Code Status used for CsvSpawner EBus notify - OnEntitiesSpawnFinished.
```
Please implicitly set such flags to *success* (or *failure*) during initialization to avoid problems in the future.

pijaro commented on Gems/CsvSpawner/Code/Include/CsvSpawner/CsvSpawnerInterface.h:
```diff
@@ -0,0 +1,86 @@
+     * SpawnStatusCode provides various status indicators for entity spawning.
+     * These flags help track whether spawning was successful, stopped, or failed.
+     */
+    enum class SpawnStatusCode : uint8_t
```
No need for `Code` postfix.

`SpawnStatus` is enough.

pijaro commented on Gems/CsvSpawner/Code/Include/CsvSpawner/CsvSpawnerInterface.h:
```diff
@@ -0,0 +1,86 @@
+    {
+        Success = 0, ///< Operation succeeded.
+        Fail = 1 << 0, ///< Generic failure.
+        SpawnStopped = 1 << 1, ///< Spawning was stopped prematurely but not necessarily a failure.
+        ErrorOccurred = 1 << 2, ///< An error occurred during spawning (potentially recoverable).
```
Keep it simpler:

```suggestion
        Stopped = 1 << 1, ///< Spawning was stopped prematurely but not necessarily a failure.
        Error = 1 << 2, ///< An error occurred during spawning (potentially recoverable).
```

pijaro commented on Gems/CsvSpawner/Code/Include/CsvSpawner/CsvSpawnerInterface.h:
```diff
@@ -0,0 +1,86 @@
+     */
+    struct SpawnInfo
+    {
+        const AZStd::vector<CsvSpawnerUtils::CsvSpawnableEntityInfo>& m_entitiesToSpawn; ///< List of entities to spawn.
+        const AZStd::string& m_physicsSceneName; ///< Name of the physics scene where entities will be spawned.
+        const AZ::EntityId& m_spawnerParentEntityId; ///< Parent entity ID managing the spawn process.
```
Why `const` references in this struct? Huge dangling reference risk here (I can imagine someone waits a frame or two after acquiring the status - the referenced objects might be gone then).

w-czerski replied:
My mistake, I've used auto-generate from Clion.

jhanca-robotecai commented:
Rethining the problem... Could we include a behavioral context for LUA/ScriptCanvas in this PR? I can imagine a script reacting to  spawned objects.

w-czerski commented:
Sure. I'll add also similar notify bus for Geo Json Spawner.

w-czerski committed:
change ros2 -> ROS 2

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
review resolved | topic changed to spline

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
review resolved | use proper ROS 2 naming

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
get ros2frame outside tick

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add update frequency

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add readme info | add update frequency

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add readme info | add update frequency

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
undo make public utils

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski commented on Gems/CsvSpawner/Code/Include/CsvSpawner/CsvSpawnerUtils.h:
Reverted.

w-czerski committed:
remove const in func definition

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
make interface clas final

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
init value definition | add spawn stop break points

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove const& from struct definition

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove final from interface

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
use constexpr

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
move struct to utils

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add lua / script canvas support

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add info to readme

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix headers | make interface func pure virtual

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
fix reflections

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk commented on Gems/CsvSpawner/Code/3rdParty/csv/csv.hpp:
revert these changes, We shouldn't apply formatting for 3rdParty code

norbertprokopiuk commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -289,7 +334,70 @@ namespace CsvSpawner::CsvSpawnerUtils
+
+        // Check is success spawn
+        tickets.empty() ? spawnStatusCode |= SpawnStatus::Fail : spawnStatusCode |= SpawnStatus::Success;
+        // Call CsvSpawner EBus notification - Finished
+        CsvSpawnerNotificationBus::Broadcast(&CsvSpawnerInterface::OnEntitiesSpawnFinished, broadcastSpawnInfo, spawnStatusCode);
```
This notification should be send when spawning actually finished, so inside completionCallback. Call in this place should happen only in case of wrong request. To be precise, it should be called when ALL entities are spawned, so in the last completion callback

w-czerski replied:
Got it! I've added a simple counter to track if all spawn occurrences are really 'spawned'. Let me know what you think.

norbertprokopiuk commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -11,14 +11,14 @@
 #include "CsvSpawnerUtils.h"
 
-#include "AzFramework/Physics/CollisionBus.h"
+#include <AzFramework/Physics/CollisionBus.h>
```
is there a reason for this header? you don't use it anyway

w-czerski replied:
It is used in lines 279-280 for `Physics::CollisionRequestBus`.
```cpp
Physics::CollisionRequestBus::BroadcastResult(
                    collisionGroup, &Physics::CollisionRequests::GetCollisionGroupById, spawnConfig.m_selectedCollisionGroup);
 ```

norbertprokopiuk replied:
My bad, I must have missed that part somehow.

michalpelka committed:
Adjust Pointcloud, ImGuiProvider, ImGuizmo code to 2505 version of o3de. (#118)

* Adjust code to 2505 version of o3de.

Co-authored-by: moudgils <47460854+moudgils@users.noreply.github.com>
Signed-off-by: Michał Pełka <michal.pelka@robotec.ai>

* Make clang happy

Signed-off-by: Michał Pełka <michal.pelka@robotec.ai>

* make ImGuiProvider work with 25050 (#121)

Signed-off-by: Mateusz Żak <mateusz.zak@robotec.ai>

* Adjust ImGuizmo Gem to compile with 2505

Signed-off-by: Michał Pełka <michal.pelka@robotec.ai>

---------

Signed-off-by: Michał Pełka <michal.pelka@robotec.ai>
Signed-off-by: Mateusz Żak <mateusz.zak@robotec.ai>
Co-authored-by: moudgils <47460854+moudgils@users.noreply.github.com>
Co-authored-by: Mateusz Żak <mateusz.zak@robotec.ai>

patrykantosz committed:
Update GeoJSONSpawner to be compatible with O3DE 2505 (#124)

Signed-off-by: Patryk Antosz <patryk.antosz@robotec.ai>

w-czerski committed:
revert 3rdParty changes

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
add counter to track if all entities are spawned

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -317,7 +321,7 @@ namespace CsvSpawner::CsvSpawnerUtils
             };
             optionalArgs.m_completionCallback =
-                [parentId, &spawnStatusCode](
+                [parentId, &spawnStatusCode, &broadcastSpawnInfo, &tickets, totalTickets, &completedTickets](
```
It appears that the four references in question are currently dangling. This arises because local variables are being passed to the completion callback, which is executed after the `SpawnEntities` function has ended. Once the function finishes, the references to the callback become invalid since the variables they point to no longer exist. It is important to address this issue to ensure the code functions as intended.

norbertprokopiuk replied:
The reference to `tickets` is not only dangling but also remains unused, which consequently interferes with building O3DE in the profile configuration.

norbertprokopiuk commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -329,16 +333,28 @@ namespace CsvSpawner::CsvSpawnerUtils
+                {
+                    // Call CsvSpawner EBus notification - Finished
+                    CsvSpawnerNotificationBus::Broadcast(&CsvSpawnerInterface::OnEntitiesSpawnFinished, broadcastSpawnInfo, spawnStatusCode);
+                }
```
Even if we assume that `completedTickets` is not a dangling reference, the current condition will not function as intended. The issue arises because `totalTicket` is passed as a copy and incremented after the callback is defined. Consequently, `completedTickets` will always be one greater than `totalTickets`, resulting in the condition never being met. If you move the `totalTickets++` before defining the callback, it would trigger with every callback. Therefore, it is advisable to explore a different approach to address this condition.

norbertprokopiuk commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -317,7 +321,7 @@ namespace CsvSpawner::CsvSpawnerUtils
+                [parentId, &spawnStatusCode, &broadcastSpawnInfo, &tickets, totalTickets, &completedTickets](
                     [[maybe_unused]] AzFramework::EntitySpawnTicket::Id ticketId, AzFramework::SpawnableConstEntityContainerView view)
             {
                 if (view.empty())
```
The code within this if statement exhibits undefined behavior because `spawnStatusCode` is a dangling reference. This may be the reason why you have not encountered any issues, as the code possibly has not executed as expected. Additionally, the code is missing a notification call to indicate a failure. A similar situation is present in the `preinsertionCallback`.

w-czerski committed:
Merge branch 'main' into wc/csvspawner_notifications

w-czerski committed:
remove unused directives

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk reviewed (changes requested):
I checked the code and I found some key issues that need your attention:

1. **Unused Variable**: The presence of an unused lambda capture relating to `tickets` is causing compilation issues. This needs to be either utilized or removed to get the code to compile.

2. **Dangling References**: There are dangling references, which can lead to unexpected behavior.

3. **Logical Issues**: The conditional logic involving `totalTicket` and `completedTickets` isn't set up correctly. This results in conditions that can never be met due to how the variables are managed.

4. **Missing Notifications**: It's important to include notification calls to alert users about any failures.

I will be happy to check refined version

w-czerski committed:
clang format

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

w-czerski committed:
remove dangling references and pointers | using shared ptr for async lamdas

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -11,6 +11,8 @@
 
 #include "CsvSpawnerUtils.h"
 
+#include "AzCore/std/smart_ptr/make_shared.h"
```
```suggestion
#include <AzCore/std/smart_ptr/make_shared.h>
```

norbertprokopiuk commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -223,11 +225,22 @@ namespace CsvSpawner::CsvSpawnerUtils
         SpawnInfo broadcastSpawnInfo =
             SpawnInfo{ entitiesToSpawn, physicsSceneName, parentId }; // Spawn Info used in CsvSpawner EBus notify.
-        SpawnStatus spawnStatusCode = SpawnStatus::Success; // Spawn Status Code used for CsvSpawner EBus notify - OnEntitiesSpawnFinished.
+        // SpawnStatus spawnStatusCode = SpawnStatus::Success; // Spawn Status Code used for CsvSpawner EBus notify -
```
I recommend to remove this code if it is commented

norbertprokopiuk commented on Gems/CsvSpawner/Code/Include/CsvSpawner/CsvSpawnerInterface.h:
```diff
@@ -34,14 +34,14 @@ namespace CsvSpawner
          * @param m_statusCode Status code indicating success, failure and warnings of the spawn.
          */
-        virtual void OnEntitiesSpawnFinished(CsvSpawnerUtils::SpawnInfo& m_spawnInfo, CsvSpawnerUtils::SpawnStatus m_statusCode) = 0;
+        virtual void OnEntitiesSpawnFinished(CsvSpawnerUtils::SpawnInfo m_spawnInfo, CsvSpawnerUtils::SpawnStatus m_statusCode) = 0;
```
Is there a reason for removing those references?

norbertprokopiuk commented on Gems/CsvSpawner/Code/Include/CsvSpawner/CsvSpawnerInterface.h:
I think that created notification methods are correct, but I've noticed that you set `AddressPolicy` as `Single`. We can have multiple CSVSpawner components on the scene. It means that we should be send notification either when all components finishes their jobs (you would need system component for that), or you can change this notification bus `AddressPolicy` to `ById` to allow each component to send their own notification. Then it would be client call to decide on which entity/component notification they want to connect with.

norbertprokopiuk commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -236,6 +249,8 @@ namespace CsvSpawner::CsvSpawnerUtils
         auto spawner = AZ::Interface<AzFramework::SpawnableEntitiesDefinition>::Get();
         AZ_Assert(spawner, "Unable to get spawnable entities definition");
 
+        auto pendingSpawns = AZStd::make_shared<AZStd::atomic_int>(static_cast<int>(entitiesToSpawn.size()));
```
Inside for loop through `entitiesToSpawn` you have some conditions which cause iteration skip without calling spawning interface. This causes lack of decremention of `pendingSpawns` what leads to situation where condition in line 349 is never met if as least one entity failed to spawn.

norbertprokopiuk commented on Gems/CsvSpawner/Code/Source/CsvSpawner/CsvSpawnerUtils.cpp:
```diff
@@ -320,40 +331,31 @@ namespace CsvSpawner::CsvSpawnerUtils
+                }
+                else
+                {
+                    const AZ::Entity* root = *view.begin();
+                    AZ::TransformBus::Event(root->GetId(), &AZ::TransformBus::Events::SetParent, parentId);
```
```suggestion
                    const AZ::Entity* root = *view.begin();
                    AZ_Assert(root, "Invalid root entity");
                    AZ::TransformBus::Event(root->GetId(), &AZ::TransformBus::Events::SetParent, parentId);
```

w-czerski committed:
add spacing

Signed-off-by: Wojciech Czerski <wojciech.czerski@robotec.ai>

norbertprokopiuk reviewed (commented):
I checked your fix and undefined behaviour are removed, which is nice. Good Job. Although there are some issues still present.
- there is invalid logic with decrementing of `pendingSpawns` 
- I think that notifcation bus definition is a bit wrong. We can add multiple CSVComponents on scene, so we should either send one notification when all components finished their jobs or each of them should send individual notification. Currently all notifications are broadcasted what is not optimal solution

Could you also explain why you decide to solve sending notification in the current way? I was thinking if wouldn't it be cleaner to define internal event, only in this Gem, which would be trigger everytime single completion callback is executed. It would take status, and cmopleted ticketId and would trigger the event on entityId of the componentOwner. Each component already has map of spawnedTickets. you would need to add unordered_set, which will gather all completedTickets inside mentioned event callback. If all tickets are completed, then you can send main notification that spawning is finished. I think it could potentially limit amount of data passed between threads and it might be a bit cleaner. But I don't want to force you to change the implementation right now. I would like to know your point of view on this problem.
